src/Windows/Core/Impl/Security/SecurityService.cs

```diff
@@ -60,23 +61,38 @@ public class SecurityService : ISecurityService {
 
         public bool DeleteUserCredentials(string authority) => NativeMethods.CredDelete(authority, NativeMethods.CRED_TYPE.GENERIC, 0);
 
-        public string GetUserName(string authority) {
+        public (string username, SecureString password) ReadUserCredentials(string authority) {
             using (var ch = CredentialHandle.ReadFromCredentialManager(authority)) {
-                if (ch != null) {
-                    var credData = ch.GetCredentialData();
-                    return credData.UserName;
+                if (ch == null) {
+                    return (null, null);
                 }
-                return string.Empty;
+
+                var credData = ch.GetCredentialData();
+                return (credData.UserName, SecurityUtilities.SecureStringFromNativeBuffer(credData.CredentialBlob));
             }
         }
 
-        private Credentials ReadSavedCredentials(string authority) {
-            using (var ch = CredentialHandle.ReadFromCredentialManager(authority)) {
-                if (ch != null) {
-                    var credData = ch.GetCredentialData();
-                    return Credentials.Create(credData.UserName, SecurityUtilities.SecureStringFromNativeBuffer(credData.CredentialBlob));
+        public void SaveUserCredentials(string authority, string userName, SecureString password, bool save) {
+            var creds = default(NativeMethods.CredentialData);
+            try {
+                creds.TargetName = authority;
+                // We have to save the credentials even if user selected NOT to save. Otherwise, user will be asked to enter
```

Comment:
This sounds insecure. Why are we not respecting user's intent not to save? Perhaps they want to put container somewhere or have encrypted disk and want data to be secure. This may not pass sec review.

Reply:
This isn't related to the container. These credentials are used by connection to that container and aren't stored in the container itself. The same logic is used when we connect to any broker using URL.

Reply:
But then it is even less secure. We allow reconnection without re-entering creds?

Reply:
https://msdn.microsoft.com/en-us/library/windows/desktop/aa374788(v=vs.85).aspx


Value | Meaning
-- | --
CRED_PERSIST_SESSION 1 (0x1) | The credential persists for the life of the logon session.  It will not be visible to other logon sessions of this same user. It  will not exist after this user logs off and back on.
CRED_PERSIST_LOCAL_MACHINE 2 (0x2) | The credential persists for all subsequent logon sessions on this  same computer. It is visible to other logon sessions of this same user  on this same computer and not visible to logon sessions for this user on  other computers. Windows Vista Home Basic, Windows Vista Home Premium, Windows Vista Starter and Windows XP Home Edition:  This value is not supported.
CRED_PERSIST_ENTERPRISE 3 (0x3) | The credential persists for all subsequent logon sessions on this  same computer. It is visible to other logon sessions of this same user  on this same computer and to logon sessions for this user on other  computers. This option can be implemented as locally persisted credential if the  administrator or user configures the user account to not have roam-able  state. For instance, if the user has no roaming profile, the credential  will only persist locally. Windows Vista Home Basic, Windows Vista Home Premium, Windows Vista Starter and Windows XP Home Edition:  This value is not supported.

Since we save it for CRED_PERSIST_SESSION, user won't be able to reuse these credentials from other sessions.

Reply:
OK, Need to study it more though...

Reply:
The case were this is iffy, is when the user says they don't want to save the creds. The problem is if we don't save it (temporarily) any connection attempt will request user to enter credentials. if the save is disabled, you will see credentials prompt show to you at least 4 times (Test connection, repl connection, intellisense, etc), before the connection is ready. Any call to broker (except About and info), will require the credentials and credential window will be presented. The iffy part is when this information is removed. Currently it is removed when user logs-out of windows . But we could remove it on connection switch, if the save flag is not set. But that will be an explicit remove.